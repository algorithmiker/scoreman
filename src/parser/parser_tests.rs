use crate::{parser::parser::parse, BufLines};

#[test]
fn test_score() {
    let example_score = r#"
e|---|
B|-3-|
G|6-6|
D|---|
A|---|
E|---|

e|---|
B|3-3|
G|-6-|
D|---|
A|---|
E|---|
"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(example_score.into())).dump_tracks());
}
#[test]
fn test_part() {
    let example_part = r#"
e|---|---|
B|-3-|3-3|
G|6-6|-6-|
D|---|---|
A|---|---|
E|---|---|"#;

    insta::assert_snapshot!(parse(&BufLines::from_string(example_part.into())).dump_tracks());
}

#[test]
fn test_left_bend_score() {
    let score = r#"
e|--12b-|
B|--3--0|
G|2-----|
D|------|
A|------|
E|------|
"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(score.into())).dump_tracks());
    let score = r#"
e|--12b-|
B|--3--4|
G|2-----|
D|------|
A|------|
E|------|
"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(score.into())).dump_tracks());
}

#[test]
fn test_right_bend_score() {
    let score = r#"
e|--12b|
B|-0--3|
G|2----|
D|-----|
A|-----|
E|-----|
"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(score.into())).dump_tracks());
}

#[test]
fn test_full_bend_score() {
    let score = r#"
e|--12b|
B|--12b|
G|2----|
D|-----|
A|-----|
E|-----|
"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(score.into())).dump_tracks());
    let score = r#"
e|--12b|
B|--12b|
G|2-12b|
D|--12b|
A|--12b|
E|--12b|
"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(score.into())).dump_tracks());
}

#[test]
fn test_bendy_score() {
    let example_part = r#"
e|--12b---12b-|
B|--3---0---3-|
G|2-----------|
D|------------|
A|------------|
E|------------|
"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(example_part.into())).dump_tracks());
    let score = r#"
e|--12b-12b-|
B|--3---12b-|
G|2-----12b-|
D|------12b-|
A|------12b-|
E|------12b-|
"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(score.into())).dump_tracks());
}
#[test]
fn test_multichar_tracks() -> anyhow::Result<()> {
    let input = r#"
e|----5--|
B|---3---|
G|10---12|
D|-------|
A|-------|
E|-------|"#;
    insta::assert_snapshot!(parse(&BufLines::from_string(input.into())).dump_tracks());
    Ok(())
}

#[test]
pub fn test_parse3() {
    use std::time::Instant;
    let example_score = r#"
e|--12-12|--12-12|--12-12|
B|3------|3------|3----11|
G|-6-3-3-|-6-3-3-|-6-3-11|
D|-------|-------|-----11|
A|-------|-------|-----11|
E|-----9-|-----9-|-----11|

// This is a comment!

e|--12-12|--12-12|
B|3------|3------|
G|-6-3-3-|-6-3-3-|
D|-------|-------|
A|-------|-------|
E|-----9-|-----9-|
"#;
    let time_parser3 = Instant::now();
    let parse3_result = parse(&BufLines::from_string(example_score.into()));
    println!("Parser3 took: {:?}", time_parser3.elapsed());
    insta::assert_snapshot!(parse3_result.dump_tracks());
    insta::assert_debug_snapshot!(parse3_result);
}
