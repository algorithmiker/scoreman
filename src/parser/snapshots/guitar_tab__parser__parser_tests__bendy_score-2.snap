---
source: src/parser/parser_tests.rs
expression: "Parser2::default().parse(score.lines())"
---
Ok(
    Parse2Result {
        diagnostics: [],
        sections: [
            Part {
                part: [
                    Partline {
                        string_name: 'e',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'B',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'G',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'D',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'A',
                        measures: 0..=0,
                    },
                    Partline {
                        string_name: 'E',
                        measures: 0..=0,
                    },
                ],
            },
        ],
        string_names: [
            'e',
            'B',
            'G',
            'D',
            'A',
            'E',
        ],
        strings: [
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: FretBend(
                        12,
                    ),
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: FretBend(
                        12,
                    ),
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Fret(
                        3,
                    ),
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: FretBend(
                        12,
                    ),
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Fret(
                        2,
                    ),
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: FretBend(
                        12,
                    ),
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: FretBend(
                        12,
                    ),
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: FretBend(
                        12,
                    ),
                },
                RawTick {
                    element: Rest,
                },
            ],
            [
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: Rest,
                },
                RawTick {
                    element: FretBend(
                        12,
                    ),
                },
                RawTick {
                    element: Rest,
                },
            ],
        ],
        measures: [
            [
                Measure {
                    content: 0..=5,
                    parent_line: 1,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=5,
                    parent_line: 2,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=5,
                    parent_line: 3,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=5,
                    parent_line: 4,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=5,
                    parent_line: 5,
                    index_on_parent_line: 1,
                },
            ],
            [
                Measure {
                    content: 0..=5,
                    parent_line: 6,
                    index_on_parent_line: 1,
                },
            ],
        ],
        offsets: [
            [
                3,
                4,
                7,
                8,
                11,
                12,
            ],
            [
                17,
                18,
                21,
                22,
                25,
                26,
            ],
            [
                31,
                32,
                35,
                36,
                39,
                40,
            ],
            [
                45,
                46,
                49,
                50,
                53,
                54,
            ],
            [
                59,
                60,
                63,
                64,
                67,
                68,
            ],
            [
                73,
                74,
                77,
                78,
                81,
                82,
            ],
        ],
        bend_targets: {},
    },
)
